### DISCUSSION TITLE
How to create a Bitmap from an image RGB buffer?

### DISCUSSION THREAD

**Discussion by pzs7602** [AvaloniaUI/Avalonia#9873] [Q&A] [answered]
In windows platform, I can create a Avalonia.Media.Imaging.Bitmap by using System.Drawing.Bitmap from an image RGB buffer:
```
        // buffer size = width*height*3
        public static Avalonia.Media.Imaging.Bitmap GetBitmapFromRGBBuffer(byte[] buffer, int width, int height)
        {
            System.Drawing.Bitmap b = new System.Drawing.Bitmap(width, height, System.Drawing.Imaging.PixelFormat.Format24bppRgb);

            System.Drawing.Rectangle BoundsRect = new System.Drawing.Rectangle();
            BoundsRect.Width = width;
            BoundsRect.Height = height;
            
            BitmapData bmpData = b.LockBits(BoundsRect, ImageLockMode.WriteOnly, b.PixelFormat);

            IntPtr ptr = bmpData.Scan0;

            // add back dummy bytes between lines, make each line be a multiple of 4 bytes
            int skipByte = bmpData.Stride - width*3;
            byte[] newBuff = new byte[buffer.Length + skipByte*height];
            for (int j = 0; j < height; j++)
            {
                Buffer.BlockCopy(buffer, j * width * 3, newBuff, j * (width * 3 + skipByte), width * 3);
            }

            // fill in rgbValues
            Marshal.Copy(newBuff, 0, ptr, newBuff.Length);
            b.UnlockBits(bmpData);
            return ConvertToAvaloniaBitmap(b);
        }
        public static Avalonia.Media.Imaging.Bitmap ConvertToAvaloniaBitmap(System.Drawing.Bitmap bitmap)
        {
            if (bitmap == null)
                return null;
            System.Drawing.Bitmap bitmapTmp = bitmap;
            var bitmapdata = bitmapTmp.LockBits(new Rectangle(0, 0, bitmapTmp.Width, bitmapTmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
            Avalonia.Media.Imaging.Bitmap bitmap1 = new Avalonia.Media.Imaging.Bitmap(Avalonia.Platform.PixelFormat.Bgra8888, Avalonia.Platform.AlphaFormat.Premul, bitmapdata.Scan0, new Avalonia.PixelSize(bitmapdata.Width, bitmapdata.Height), new Avalonia.Vector(96, 96), bitmapdata.Stride);
            bitmapTmp.UnlockBits(bitmapdata);
            bitmapTmp.Dispose();
            return bitmap1;
        }

```
Now, I want to implement the method GetBitmapFromRGBBuffer in macOS/Linux, is it possible to do this directly in Avalonia?(System.Drawing.Bitmap is not available in macOS/Linux).
Thanks!

**Comment by maxkatz6** [comment, marked as answer]
You can use WriteableBitmap and WriteableBitmap.Lock APIs instead of System.Drawing.Bitmap.
WriteableBitmap API operates with raw pointer to which you can write your data. 

If you need more highlevel API for drawing, you can use third party libraries like SkiaSharp or ImageSharp and convert them to avalonia bitmap similarly how you did it in ConvertToAvaloniaBitmap.

**Reply by robloo** [reply]
Take a look at the ColorPicker code too if you want. It generated bitmaps from calculated byte arrays.

https://github.com/AvaloniaUI/Avalonia/blob/228f26f0a1164cbd56efe58ef86525193709d8bb/src/Avalonia.Controls.ColorPicker/Helpers/ColorPickerHelpers.cs#L606-L626

**Reply by pzs7602** [reply]
Thanks for the reply, I try to modify the code to adapt to my environment:
```
        public static WriteableBitmap CreateBitmapFromPixelData(byte[] rgbPixelData, int width, int height)
        {
            // Standard may need to change on some devices
            Vector dpi = new Vector(96, 96);
            // the image has no alpha channel
            var bitmap = new WriteableBitmap(new PixelSize(width, height), dpi, Avalonia.Platform.PixelFormat.Rgb565);
            using (var frameBuffer = bitmap.Lock())
            {
                Marshal.Copy(rgbPixelData, 0, frameBuffer.Address, rgbPixelData.Length);
            }
            
            return bitmap;
        }

```
My environment :
```
rgbPixelData is the image RGB data buffer(length = width*height*3)
width, height are the bitmap width and height(in pixel)
the image has no alpha channel
```
When the method get called, my app is crashed(debug message is EXC_BAD_ACCESS (SIGSEGV)). What's the problem? is the usage: new PixelSize(width, height) or Avalonia.Platform.PixelFormat.Rgb565 correct?

**Reply by maxkatz6** [reply]
In your previous code you used Format24bppRgb which is (from WPF doc) - "Specifies that pixel format is 24 bits per pixel. The color information specifies 16777216 shades of color of which 8 bits are red, 8 bits are green and 8 bits are blue.".

In Avalonia code you are using Rgb565 which combines two pixels into two bytes (5 for red, 6 for green, 5 for blue).

The closest you need in Avalonia is Rgba8888 (or Bgra8888), but it also requires alpha byte. Which means you would need to adapt your code or convert bytes.

**Reply by pzs7602** [reply]
Thanks! @maxkatz6

**Reply by ahopper** [reply]
you really should respect WriteableBitmap.RowBytes, there is no guarantee that the stride of the bitmap is the same as the width.

**Comment by dorisoy** [comment]
> In windows platform, I can create a Avalonia.Media.Imaging.Bitmap by using System.Drawing.Bitmap from an image RGB buffer:
> 
> ```
>         // buffer size = width*height*3
>         public static Avalonia.Media.Imaging.Bitmap GetBitmapFromRGBBuffer(byte[] buffer, int width, int height)
>         {
>             System.Drawing.Bitmap b = new System.Drawing.Bitmap(width, height, System.Drawing.Imaging.PixelFormat.Format24bppRgb);
> 
>             System.Drawing.Rectangle BoundsRect = new System.Drawing.Rectangle();
>             BoundsRect.Width = width;
>             BoundsRect.Height = height;
>             
>             BitmapData bmpData = b.LockBits(BoundsRect, ImageLockMode.WriteOnly, b.PixelFormat);
> 
>             IntPtr ptr = bmpData.Scan0;
> 
>             // add back dummy bytes between lines, make each line be a multiple of 4 bytes
>             int skipByte = bmpData.Stride - width*3;
>             byte[] newBuff = new byte[buffer.Length + skipByte*height];
>             for (int j = 0; j < height; j++)
>             {
>                 Buffer.BlockCopy(buffer, j * width * 3, newBuff, j * (width * 3 + skipByte), width * 3);
>             }
> 
>             // fill in rgbValues
>             Marshal.Copy(newBuff, 0, ptr, newBuff.Length);
>             b.UnlockBits(bmpData);
>             return ConvertToAvaloniaBitmap(b);
>         }
>         public static Avalonia.Media.Imaging.Bitmap ConvertToAvaloniaBitmap(System.Drawing.Bitmap bitmap)
>         {
>             if (bitmap == null)
>                 return null;
>             System.Drawing.Bitmap bitmapTmp = bitmap;
>             var bitmapdata = bitmapTmp.LockBits(new Rectangle(0, 0, bitmapTmp.Width, bitmapTmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
>             Avalonia.Media.Imaging.Bitmap bitmap1 = new Avalonia.Media.Imaging.Bitmap(Avalonia.Platform.PixelFormat.Bgra8888, Avalonia.Platform.AlphaFormat.Premul, bitmapdata.Scan0, new Avalonia.PixelSize(bitmapdata.Width, bitmapdata.Height), new Avalonia.Vector(96, 96), bitmapdata.Stride);
>             bitmapTmp.UnlockBits(bitmapdata);
>             bitmapTmp.Dispose();
>             return bitmap1;
>         }
> ```
> 
> Now, I want to implement the method GetBitmapFromRGBBuffer in macOS/Linux, is it possible to do this directly in Avalonia?(System.Drawing.Bitmap is not available in macOS/Linux). Thanks!


If your sourcePtr is a pointer IntPtr to the image pixel buffer, you can write it this way：

```

/// <summary>
/// 将IntPtr（指向图像像素缓冲区的指针）复制到可写入位图
/// </summary>
/// <param name="writeableBitmap"></param>
/// <param name="sourcePtr"></param>
/// <param name="width"></param>
/// <param name="height"></param>
/// <param name="stride"></param>
public static void CopyIntPtrToWriteableBitmap(WriteableBitmap writeableBitmap, IntPtr sourcePtr, int width, int height, int stride)
{
    // 锁定 WriteableBitmap，准备写入数据
    using (var lockedBuffer = writeableBitmap.Lock())
    {
        // 遍历图片的每一行
        for (int y = 0; y < height; y++)
        {
            // 创建一个字节数组来存储这一行的数据
            byte[] rowData = new byte[width * 4];// 假设是32位位图，每个像素4字节

            // 源指针偏移，定位到当前行的起始位置
            IntPtr srcPtrOffset = IntPtr.Add(sourcePtr, y * stride);

            // 从非托管内存复制到托管数组
            Marshal.Copy(srcPtrOffset, rowData, 0, rowData.Length);

            // 将托管数组的内容复制到WriteableBitmap
            for (int i = 0; i < rowData.Length; i++)
            {
                // 计算目标位置指针
                IntPtr destPtrOffset = IntPtr.Add(lockedBuffer.Address, y * lockedBuffer.RowBytes + i);
                // 写入数据
                Marshal.WriteByte(destPtrOffset, rowData[i]);
            }
        }
    }
}

```
